File: placementprepration/tree/checkbalance.cpp
#include<iostream>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        this->data = val;
        this->left = NULL;
        this->right = NULL;
    }
};

Node* BuildTree(Node* root) {
    int data;
    cout << "Enter the data: ";
    cin >> data;

    if (data == -1) {
        return NULL; 
    }

    root = new Node(data);

    cout << "Enter the data in left of " << data << endl;
    root->left = BuildTree(root->left);
    cout << "Enter the data in right of " << data << endl;
    root->right = BuildTree(root->right);

    return root;
}

int height(Node* root, bool &balanced) {
    if (root == NULL) {
        return 0; 
    }

    int lh = height(root->left, balanced);
    int rh = height(root->right, balanced);

    if (abs(lh - rh) > 1) {
        balanced = false;
    }

    return max(lh, rh) + 1;
}

bool isBalanced(Node* root) {
    bool balanced = true;
    height(root, balanced);
    return balanced;
}

int main() {
    Node* root = NULL;
    root = BuildTree(root);

    if (isBalanced(root)) {
        cout << "The tree is balanced." << endl;
    } else {
        cout << "The tree is NOT balanced." << endl;
    }
    return 0;
}

File: placementprepration/tree/checkbalance.py
import sys

class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


def is_balanced(root):
    def height(node):
        if not node:
            return 0, True
        lh, lb = height(node.left)
        rh, rb = height(node.right)
        balanced = lb and rb and abs(lh - rh) <= 1
        return max(lh, rh) + 1, balanced
    return height(root)[1]


def main():
    # Placeholder tree
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    print("The tree is balanced." if is_balanced(root) else "The tree is NOT balanced.")


if __name__ == "__main__":
    main()
